### DISCUSSION TITLE
libtiff: take advantage of new TIFFClientOpenExt / TIFFOpenOptions feature

### DISCUSSION THREAD

**Discussion by lovell** [libvips/libvips#3892] [Ideas]
The example TIFF image found via fuzz testing at https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=65137 contains thousands of invalid tags, which libtiff quite happily attempts to allocate memory for. In this example a <1MB file consumes >3GB of RAM before failing.

Luckily recent versions of libtiff provide an *extended* `TIFFClientOpenExt` API (to eventually replace `TIFFClientOpen`) that allows us to pass in new `TIFFOpenOptions`.

Arguably the most useful of these is the does-what-it-says-on-the-tin `TIFFOpenOptionsSetMaxCumulatedMemAlloc`, which is about to be made available from libtiff v4.6.1 onwards.

We should probably take advantage of this feature when available. This discussion is to determine the best way to expose this to libvips' users.

Possible options (there will be others):

- Add the now-common `unlimited` flag to `tiffload` that when set to `false` (the default), uses this new API to set an arbitrary (libvips set and owned) RAM limit.
- Add a new format-specific `max_cumulated_mem` (or similar) flag to `tiffload` that exposes the integral value to pass to this new API.

I'm happy to work on this, but not quite sure of the best approach.

**Comment by jcupitt** [comment]
Hi @lovell, thanks for letting us know about this.

I suppose `unlimited` sounds simplest. We could even move it in to the base loader class.

We should support `TIFFOpenOptionsSetErrorHandlerExtR` too, right now we can potentially crash if someone else steals the libtiff error handler.
